### ISSUE TITLE
Canvas animation

### ISSUE THREAD

**Issue by chazsolo** [elfalem/Leaflet.curve#16]
Do you have an idea of what it would take to implement support for canvas animation? I ask as I know it will eventually come up in the fear future and I'd like to have a reasonable response as to what effort would be involved. I'd be happy to help implement it.

For my specific case, it would be updating the offset of a dashed line to indicate movement from one coordinate to another.

**Comment by elfalem** [comment]
I'm thinking about this but I don't have any ideas currently.

For SVG rendering, [Element.animate()](https://developer.mozilla.org/en-US/docs/Web/API/Element/animate) does the heavy lifting. We just pass the options supplied by the user. Ideally the options currently passed to Element.animate() can somehow be translated to commands suitable for the Canvas renderer. I'm still trying to understand [how animations work](https://developer.mozilla.org/en-US/docs/Web/API/Canvas_API/Tutorial/Basic_animations) in Canvas to determine if this is plausible.

**Comment by chazsolo** [comment]
I'm not an expert on canvas animations, nor is my team, but we'd be willing to help if you need it. We recently found [Canvas Flowmap Layer](https://github.com/jwasilgeo/Leaflet.Canvas-Flowmap-Layer/blob/master/src/CanvasFlowmapLayer.js), so it's definitely possible. Since your library is narrow in scope I think it may be relatively simple to add a few canned animations with options (instead of dynamic user-driven ones). I'll keep at it and let you know if I come up with anything!

**Comment by elfalem** [comment]
Thanks for referring to Canvas Flowmap Layer. That was the breakthrough I needed. I've done an implementation of animations in canvas on [this branch](https://github.com/elfalem/Leaflet.curve/tree/issue-16-canvas-animation). I haven't published to npm yet.

According to my testing, I think it works for your use case of updating the offset of a dashed line. If you'e able to test, I'd appreciate any feedback.

**Comment by chazsolo** [comment]
Thanks @elfalem, we'll give it a shot this week sometime and I'll keep you updated.

**Comment by jeffshaver** [comment]
@elfalem, @chazsolo is out this week. so I am taking a look. I got the lines to be dashed (using `dashArray`), but when i add the animate option, nothing changes. I took a  line from the examples by adding `animate: {duration: 3000, iterations: Infinity}` to my options object, to no avail. any help?

to add as much info as possible, heres the entire call to Leaflet.curve:

```javascript
const curve = L.curve(['M', latLng, 'Q', midPointLatLng, oLatLng], {
      color: theme.palette.slate.light,
      opacity: ghost ? 0.2 : 1,
      // TODO render Arcs, countries, and other elements in the same canvas?
      pane: PANES.Arcs,
      stroke: true,
      weight: 1.4,
      dashArray: '5',
      animate: { duration: 500, iterations: Infinity },
    })
```

**Comment by jeffshaver** [comment]
@elfalem nvm. I got it working. I didn't have `window.TWEEN` defined. 

Now the issue I am having is that some curves are animating much faster than others.

**Comment by jeffshaver** [comment]
@elfalem I think I need to figure out a way to normalize the duration of the animation regardless of distance.....

**Comment by elfalem** [comment]
@jeffshaver if the animations have the same duration, they should start and stop at the same time. If one curve is significantly longer than another curve but they have the same duration, the first one will appear to be faster since it has to cover more distance in the same time period. If it's using infinite iterations, playing around with the duration might help.

**Comment by jeffshaver** [comment]
@elfalem yeah. So, I tested the branch. the animations work (at least the one we wanted to do). and didn't require much change on our end (thanks!). We need to play around with normalizing the duration based on the distances since we can have large differences between our longest and shortest lines. 

The work you have done here has been awesome and we appreciate the help with this!

The only thing I didn't see in the docs (but I could have missed it) is that `@tweenjs/tween.js` is an optional dependency (because you may not need it if using svg I assume). That part is fine, but it would be nice if it was documented that you need to assign window.TWEEN yourself.

Another option would be to require us to provide the tweening function ourselves?

that way I could do something like

```javascript
import TWEEN from '@tweenjs/tween.js'
import { provideTween } from '@elfalem/Leaflet.curve'

provideTween(TWEEN)
```

instead of 

```javascript
import TWEEN from '@tweenjs/tween.js'

window.TWEEN = TWEEN
```

This becomes more annoying for us since we are using TypeScript. Where window actually is typed and we need to type the window.TWEEN property. If it was a function from your package that let us provide the TWEEN function, than we wouldn't need to assign window.TWEEN. We also wouldn't have to make the call of where to put that assignment. Do I put `window.TWEEN = TWEEN` in app.js? In the module that is actually using the Leaflet.curve library?

Just some thoughts. Or if you have another idea on how to do the window.TWEEN thing without specifically putting TWEEN on window?

Again, really appreciate the help you have provided!

**Comment by elfalem** [comment]
Ok, happy to hear it's working. Regarding the docs, there is a change in the README that mentions tween.js but it's on this issue branch. Once the branch is merged, it should be more visible.

I will look into a better way of providing a reference to TWEEN. It looks for window.TWEEN which is not ideal in a Node environment. Thanks for testing and the feedback!

**Comment by elfalem** [comment]
@jeffshaver I just pushed a change that removes the reference to `window` to the branch. It was used only to check for existence of the TWEEN object, so I replaced it with a `typeof` check.

**Comment by elfalem** [comment]
@chazsolo @jeffshaver any update on whether the `window` reference issue is resolved? I'm hoping to merge this and publish a release soon.

**Comment by jeffshaver** [comment]
@elfalem, @chazsolo or I will give this a shot in the next few days. Sorry for the late response!

**Comment by elfalem** [comment]
Merged and published v0.4.0
